addons/web/static/src/js/model.js

```diff
@@ -417,10 +424,42 @@ odoo.define("web/static/src/js/model.js", function (require) {
          * @returns {Promise}
          */
         async load() {
-            for (const extension of this.extensionsToLoad) {
-                await extension.load();
+            const toLoad = new Map();
+            const loaded = [];
+            // Determine which extensions need to be loaded
+            for (const extension of this.extensions) {
+                const { shouldLoad, dependencies } = extension.getLoadingState();
+                if (shouldLoad) {
+                    for (const dep of dependencies) {
+                        if (!this.extensions.some((e) => e.name === dep)) {
+                            // Any dependency not found in the registered
+                            // extensions is considered as loaded.
+                            loaded.push(dep);
+                        }
+                    }
+                    // Extension will be loaded
+                    toLoad.set(extension.name, extension);
+                } else {
+                    // Extension will not be loaded
+                    loaded.push(extension.name);
+                }
+            }
+            // Iterate on each group of extensions ready to load
+            while (toLoad.size) {
+                const loading = new Map();
+                for (const [name, extension] of toLoad) {
+                    const { dependencies } = extension.getLoadingState();
+                    // A extension can be loaded iff each of its dependency
+                    // is already loaded.
+                    if (dependencies.every((dep) => loaded.includes(dep))) {
+                        loading.set(name, extension.load());
+                        toLoad.delete(name);
+                    }
+                }
+                // Wait for the current group to be loaded
+                await Promise.all(loading.values());
```

Comment:
By doing this, if you have a tree

A -- B
C --D
meaning B depends on A and D depends on C, you will first load A, C and wait A and C to be loaded before going on.
If A takes less time than C to load, you will wait to load B instead of loading it immediately

Reply:
As discussed, the current disposition of the models and extensions does not require a more advanced dependency system (the only requirement is for all model extensions to be loaded before the SearchPanelME, thus this strategy)